src/adr/0036-against-use-of-conditional-statements-in-guice-modules.md

```diff
@@ -0,0 +1,110 @@
+ localFiles and LinShare.
+ - Text extraction: Extracting text from attachment to enable attachment search. There is a Tika implementation, but 
+ lighter JSOUP based options are also available.
+
+In order to keep the cardinality of Guice products low, we decided to use conditional statements in modules based on the 
+configuration to select which one to enable. Eventually defeating the Guice adoption goals mentioned above.
+
+Finally, Blob Storing technology offers a wide combination of technologies:
+
+ - ObjectStorage in itself could implement either Swift APIs or Amazon S3 APIs
+ - We decided to keep supporting Cassandra for blob storing as an upgrade solution from Cassandra-guice to Distributed 
+James for existing users.
+ - Proposals such as [HybridBlobStore](0014-blobstore-storage-policies.md) and then 
+[Cassandra BlobStore cache](0025-cassandra-blob-store-cache.md) proposed to leverage Cassandra as a performance 
+(latency) enhancer for ObjectStorage technologies.
+
+Yet again it had been decided to use conditional statements in modules in order to lower the cardinality.
+
+However, [Cassandra BlobStore cache](0025-cassandra-blob-store-cache.md) requires expensive resource initialization
+requiring to perform upgrade procedure (usage of an additional cache keyspace) that represents a cost we don't want to
+pay if we don't rely on that cache. Not having the cache module thus enables quickly auditing that the caching cassandra 
+session is not initialized. See 
+[this comment](https://github.com/linagora/james-project/pull/3261#pullrequestreview-389804841) as well as 
+[this comment](https://github.com/linagora/james-project/pull/3261#issuecomment-613911695).
+
+### Audit
+
+The following modules perform conditional statements upon injection time:
+
+ - BlobExportMechanismModule : Choice of the export mechanism
+ - ObjectStorageDependenciesModule::selectBlobStoreBuilder: Choice between S3 and Swift ObjectStorage technologies
+ - TikaMailboxModule::provideTextExtractor: Choice of text extraction technology
+ - BlobStoreChoosingModule::provideBlobStore: Choice of BlobStore technology: Cassandra, ObjectStorage or Hybrid
+ - [Cached blob store](https://github.com/linagora/james-project/pull/3319) represents a similar problem: should the 
+ blobStore be wrapped by a caching layer?
+
+## Decision
+
+We should no longer rely on conditional statements in Guice module.
+
+We should within James main method, upon James startup read the configuration and select the modules that should be 
+selected to run it, before calling the Guice injector to perform its full startup.
+
+This enables easy diagnose of the running components via the selected module list. It exposes tested, safe choices to 
+the user while limiting the Guice products count.
+
+Basic minimalistic integration tests will be written to cover the possibilities exposed to the user, by statically 
```

Comment:
Ok, so we have 6 products, 2 blob exports, 3 object storages, 2 content extractors, it means we have 72 (6*2*3*2) combinations to tests?
It looks doable but the problem is: it doesn't scale as each option doubles the number of tests, whether it's done with guice-less condition or not.
Do we agree?

Reply:
Idem